### PR TITLE
Add option to honor_timestamps

### PR DESCRIPTION
**Pull request by paulfantom**
Adding an option to allow choosing if prometheus should honor timestamps of incoming metrics or not.

I have also extended unit tests for prometheus configuration to cover `global` section.

/cc @LiliC @s-urbaniak

### CODE REVIEW

File: pkg/prometheus/promcfg_test.go [hunk 1]
```diff
@@ -50,6 +50,144 @@ func TestConfigGeneration(t *testing.T) {
...
+func TestGlobalSettings(t *testing.T) {
+	type testCase struct {
+		EvaluationInterval string
+		ScrapeInterval     string
+		HonorTimestamps    *bool
+		ExternalLabels     map[string]string
+		Expected           string
+	}
+
+	//honor := true
```
Comment (LiliC) [inline pkg/prometheus/promcfg_test.go]: Is this leftover from testing?

File: pkg/prometheus/promcfg_test.go [hunk 2]
```diff
@@ -50,6 +50,144 @@ func TestConfigGeneration(t *testing.T) {
...
+			nil,
+			nil,
+			nil,
+			nil,
+		)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if tc.Expected != string(cfg) {
+			t.Fatalf("Unexpected result.\n\nGot:\n\n%s\n\nExpected:\n\n%s\n\n", string(cfg), tc.Expected)
```
Comment (LiliC) [inline pkg/prometheus/promcfg_test.go]: Hmm did you mean Got instead of Unexpected?

File: Documentation/api.md [hunk 3]
```diff
@@ -394,6 +394,7 @@ PrometheusSpec is a specification of the desired behavior of the Prometheus clus
 | priorityClassName | Priority class assigned to the Pods | string | false |
 | portName | Port name used for the pods and governing service. This defaults to web | string | false |
 | arbitraryFSAccessThroughSMs | ArbitraryFSAccessThroughSMs configures whether configuration based on a service monitor can access arbitrary files on the file system of the Prometheus container e.g. bearer token files. | [ArbitraryFSAccessThroughSMsConfig](#arbitraryfsaccessthroughsmsconfig) | false |
+| honorTimestamps | HonorTimestamps configures prometheuses to honor timestamps coming from metrics | *bool | false |
```
Comment (LiliC) [inline Documentation/api.md]:
```suggestion
| honorTimestamps | HonorTimestamps configures Prometheus to honor timestamps coming from metrics | *bool | false |
```

I would rephrase this description because this is configurable. Something like:
> ... configures whether should honor timestamps....

File: pkg/prometheus/promcfg_test.go (hunk 2 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: No, this is ok. It returns something like this when failing:
```
    promcfg_test.go:186: Unexpected result.
        
        Got:
        
        global:
          evaluation_interval: 30s
          scrape_interval: 30s
          honor_timestamps: true
          external_labels:
            key1: value1
            key2: value2
            prometheus: /
            prometheus_replica: $(POD_NAME)
        rule_files: []
        scrape_configs: []
        alerting:
          alert_relabel_configs:
          - action: labeldrop
            regex: prometheus_replica
          alertmanagers: []
        
        
        Expected:
        
        global:
          evaluation_interval: 30s
          scrape_interval: 30s
          honor_timestamps: false
          external_labels:
            key1: value1
            key2: value2
            prometheus: /
            prometheus_replica: $(POD_NAME)
        rule_files: []
        scrape_configs: []
        alerting:
          alert_relabel_configs:
          - action: labeldrop
            regex: prometheus_replica
          alertmanagers: []

```

File: pkg/prometheus/promcfg_test.go (hunk 1 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: yup, fixed :)

**Comment by brancz** [comment]
I’m a little worried about the strategy here as when we/if we introduce the ability to set this on a per servicemonitor/podmonitor basis we won’t be enforcing this anymore and the whole point of this is that it’s a protection mechanism.

**Comment by s-urbaniak** [comment]
Agreed, this should be command line option to prometheus operator, no?

**Comment by brancz** [comment]
It's ok to be an option in the prometheus object I think, but we need to make sure that the semantic is that whatever we set here is _enforced_. So that when we introduce the ability to configure this on ServiceMonitor/PodMonitor level we don't have a regression.

File: Documentation/api.md [hunk 4]
```diff
@@ -394,6 +394,7 @@ PrometheusSpec is a specification of the desired behavior of the Prometheus clus
 | priorityClassName | Priority class assigned to the Pods | string | false |
 | portName | Port name used for the pods and governing service. This defaults to web | string | false |
 | arbitraryFSAccessThroughSMs | ArbitraryFSAccessThroughSMs configures whether configuration based on a service monitor can access arbitrary files on the file system of the Prometheus container e.g. bearer token files. | [ArbitraryFSAccessThroughSMsConfig](#arbitraryfsaccessthroughsmsconfig) | false |
+| honorTimestamps | HonorTimestamps configures Prometheus configures whether should honor timestamps coming from metrics | *bool | false |
```
Comment (LiliC) [inline Documentation/api.md]: one configures too many? :)

File: pkg/prometheus/promcfg_test.go (hunk 2 above)
Comment (LiliC) [reply pkg/prometheus/promcfg_test.go]: Yeah I noticed we seem to do this everywhere, was just confusing to read.

File: Documentation/api.md (hunk 4 above)
Comment (paulfantom) [reply Documentation/api.md]: fixed :)

**Comment by paulfantom** [comment]
Included code review suggestions.

File: pkg/prometheus/promcfg_test.go [hunk 5]
```diff
@@ -50,6 +50,120 @@ func TestConfigGeneration(t *testing.T) {
 	}
 }
 
+func TestGlobalSettings(t *testing.T) {
```
Comment (LiliC) [inline pkg/prometheus/promcfg_test.go]: What is the GlobalSettings test?

File: pkg/prometheus/promcfg_test.go [hunk 6]
```diff
@@ -523,6 +749,7 @@ rule_files: []
 scrape_configs:
 - job_name: default/testservicemonitor1/0
   honor_labels: false
+  honor_timestamps: true
```
Comment (LiliC) [inline pkg/prometheus/promcfg_test.go]: Hmm do we want to always set this always? I understood only when the user explicitly sets it? cc @brancz

File: Documentation/api.md [hunk 7]
```diff
@@ -394,6 +394,7 @@ PrometheusSpec is a specification of the desired behavior of the Prometheus clus
 | priorityClassName | Priority class assigned to the Pods | string | false |
 | portName | Port name used for the pods and governing service. This defaults to web | string | false |
 | arbitraryFSAccessThroughSMs | ArbitraryFSAccessThroughSMs configures whether configuration based on a service monitor can access arbitrary files on the file system of the Prometheus container e.g. bearer token files. | [ArbitraryFSAccessThroughSMsConfig](#arbitraryfsaccessthroughsmsconfig) | false |
+| honorTimestamps | HonorTimestamps configures Prometheus whether should honor timestamps coming from metrics | *bool | false |
```
Comment (LiliC) [inline Documentation/api.md]:
```suggestion
| honorTimestamps | HonorTimestamps configures Prometheus whether it should honor timestamps coming from metrics | *bool | false |
```

File: pkg/prometheus/promcfg_test.go (hunk 5 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: It is testing the prometheus parameters described below and was introduced as part of testing a previous way of setting honor_timestamps. Now I can move it to other PR or leave it here:
```
evaluation_interval
scrape_interval
external_labels
```

File: pkg/prometheus/promcfg_test.go (hunk 6 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: Currently, the user doesn't have a way of setting `honor_timestamps` option and this way it is always enforced and clear what is happening. Also `true` is a default for prometheus.

**Comment by s-urbaniak** [comment]
@paulfantom is this good for a final pass?

**Comment by paulfantom** [comment]
@s-urbaniak yes, it's ready

File: pkg/prometheus/promcfg_test.go (hunk 6 above)
Comment (brancz) [reply pkg/prometheus/promcfg_test.go]: This seems right, under the premise that potentially we want to enable people to set this on a per servicemonitor/podmonitor basis.

File: Documentation/api.md [hunk 8]
```diff
@@ -394,6 +394,7 @@ PrometheusSpec is a specification of the desired behavior of the Prometheus clus
 | priorityClassName | Priority class assigned to the Pods | string | false |
 | portName | Port name used for the pods and governing service. This defaults to web | string | false |
 | arbitraryFSAccessThroughSMs | ArbitraryFSAccessThroughSMs configures whether configuration based on a service monitor can access arbitrary files on the file system of the Prometheus container e.g. bearer token files. | [ArbitraryFSAccessThroughSMsConfig](#arbitraryfsaccessthroughsmsconfig) | false |
+| honorTimestamps | HonorTimestamps configures Prometheus whether it should honor timestamps coming from metrics | *bool | false |
```
Comment (LiliC) [inline Documentation/api.md]: This reads a bit strange to me. It implies we always set the `honor_timestamps` to false, which is not true because the default value is `true`.

WDYT of naming this `doNotHonorTimestamps` or something similar, then false will make sense to be the default value? Its an awful naming, but hopefully it gets the point across.

Comment (brancz) [reply Documentation/api.md]: Maybe we should implement the per servicemonitor/podmonitor setting in this PR and have this named as a parallel to #2806 ?

**Comment by paulfantom** [comment]
I rewrote whole thing to implement honor_timestamps per PodMonitor/ServiceMonitor and allow overriding this option from PrometheusSpec. This is now similar to how honor_labels work.

Also testing global settings was split into separate PR at https://github.com/coreos/prometheus-operator/pull/2811

/cc @brancz @LiliC

File: Documentation/api.md [hunk 9]
```diff
@@ -204,6 +204,7 @@ Endpoint defines a scrapeable endpoint serving Prometheus metrics.
 | bearerTokenFile | File to read bearer token for scraping targets. | string | false |
 | bearerTokenSecret | Secret to mount to read bearer token for scraping targets. The secret needs to be in the same namespace as the service monitor and accessible by the Prometheus Operator. | [v1.SecretKeySelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#secretkeyselector-v1-core) | false |
 | honorLabels | HonorLabels chooses the metric's labels on collisions with target labels. | bool | false |
+| honorTimestamps | HonorTimestamps allows to disable honoring timestamps | *bool | false |
```
Comment (LiliC) [inline Documentation/api.md]: But honor_timestamps should be true by default, why is this false?

File: Documentation/api.md [hunk 10]
```diff
@@ -395,6 +397,7 @@ PrometheusSpec is a specification of the desired behavior of the Prometheus clus
 | portName | Port name used for the pods and governing service. This defaults to web | string | false |
 | arbitraryFSAccessThroughSMs | ArbitraryFSAccessThroughSMs configures whether configuration based on a service monitor can access arbitrary files on the file system of the Prometheus container e.g. bearer token files. | [ArbitraryFSAccessThroughSMsConfig](#arbitraryfsaccessthroughsmsconfig) | false |
 | overrideHonorLabels | OverrideHonorLabels if set to true overrides all user configured honor_labels. If HonorLabels is set in ServiceMonitor or PodMonitor to true, this overrides honor_labels to false. | bool | false |
+| overrideHonorTimestamps | HonorTimestamps configures Prometheus whether it should honor timestamps coming from metrics | bool | false |
```
Comment (bwplotka) [inline Documentation/api.md]:
```suggestion
| overrideHonorTimestamps | OverrideHonorTimestamps configures Prometheus whether it should honor timestamps coming from metrics | bool | false |
```
?

File: pkg/apis/monitoring/v1/types.go [hunk 11]
```diff
@@ -577,6 +579,8 @@ type Endpoint struct {
 	BearerTokenSecret v1.SecretKeySelector `json:"bearerTokenSecret,omitempty"`
 	// HonorLabels chooses the metric's labels on collisions with target labels.
 	HonorLabels bool `json:"honorLabels,omitempty"`
+	// HonorTimestamps allows to disable honoring timestamps
```
Comment (bwplotka) [inline pkg/apis/monitoring/v1/types.go]: Missing trailing periods at the end of comments.

File: pkg/apis/monitoring/v1/types.go [hunk 12]
```diff
@@ -288,6 +288,8 @@ type PrometheusSpec struct {
 	// OverrideHonorLabels if set to true overrides all user configured honor_labels.
 	// If HonorLabels is set in ServiceMonitor or PodMonitor to true, this overrides honor_labels to false.
 	OverrideHonorLabels bool `json:"overrideHonorLabels,omitempty"`
+	// HonorTimestamps configures Prometheus whether it should honor timestamps coming from metrics
```
Comment (bwplotka) [inline pkg/apis/monitoring/v1/types.go]:
```suggestion
	// OverrideHonorTimestamps configures Prometheus whether it should honor timestamps coming from metrics.
```

File: pkg/apis/monitoring/v1/types.go [hunk 13]
```diff
@@ -638,6 +642,8 @@ type PodMetricsEndpoint struct {
 	ScrapeTimeout string `json:"scrapeTimeout,omitempty"`
 	// HonorLabels chooses the metric's labels on collisions with target labels.
 	HonorLabels bool `json:"honorLabels,omitempty"`
+	// HonorTimestamps allows to disable honoring timestamps
```
Comment (bwplotka) [inline pkg/apis/monitoring/v1/types.go]: ditto

File: pkg/prometheus/promcfg.go [hunk 14]
```diff
@@ -322,14 +322,29 @@ func honorLabels(userHonorLabels, overrideHonorLabels bool) bool {
 	return userHonorLabels
 }
 
+// honorTimestamps determines the value of honor_timestamps.
+// if overrideHonorTimestamps is true, then we want to always
+// set honor_timestamps to true. Otherwise we set it to whateever
```
Comment (bwplotka) [inline pkg/prometheus/promcfg.go]:
```suggestion
// set honor_timestamps to true. Otherwise, we set it to whatever
```

File: pkg/prometheus/promcfg.go [hunk 15]
```diff
@@ -322,14 +322,29 @@ func honorLabels(userHonorLabels, overrideHonorLabels bool) bool {
...
+func honorTimestamps(userHonorTimestamps *bool, overrideHonorTimestamps bool) bool {
+	honor := true
+	if userHonorTimestamps != nil {
+		honor = *userHonorTimestamps
+	}
+
+	return honor && !overrideHonorTimestamps
+}
+
 func (cg *configGenerator) generatePodMonitorConfig(version semver.Version,
```
Comment (bwplotka) [inline pkg/prometheus/promcfg.go]: this is not really Go style for large number args, but if that's kept in this codebase, then ok (: 

Should be:
```
function(
  arg1,
  arg2,
  ...
  argZ,
)
```

File: Documentation/api.md (hunk 9 above)
Comment (bwplotka) [reply Documentation/api.md]: +1 Prometheus has true by default...

File: pkg/apis/monitoring/v1/openapi_generated.go [hunk 16]
```diff
@@ -2020,6 +2034,13 @@ func schema_pkg_apis_monitoring_v1_PrometheusSpec(ref common.ReferenceCallback)
 							Format:      "",
 						},
 					},
+					"overrideHonorTimestamps": {
+						SchemaProps: spec.SchemaProps{
+							Description: "HonorTimestamps configures Prometheus whether it should honor timestamps coming from metrics",
```
Comment (bwplotka) [inline pkg/apis/monitoring/v1/openapi_generated.go]: overrideHonorTimestamps ?

File: pkg/prometheus/promcfg_test.go [hunk 17]
```diff
@@ -1581,3 +1587,53 @@ func TestHonorLabels(t *testing.T) {
...
+			OverrideHonorTimestamps: false,
+			Expected:                true,
+		},
+		{
+			UserHonorTimestamps:     &no,
+			OverrideHonorTimestamps: true,
+			Expected:                false,
+		},
+		{
+			UserHonorTimestamps:     &no,
```
Comment (bwplotka) [inline pkg/prometheus/promcfg_test.go]: `swag.Bool(..)`?

File: example/prometheus-operator-crd/prometheus.crd.yaml [hunk 18]
```diff
@@ -2831,6 +2831,10 @@ spec:
                 honor_labels. If HonorLabels is set in ServiceMonitor or PodMonitor
                 to true, this overrides honor_labels to false.
               type: boolean
+            overrideHonorTimestamps:
+              description: HonorTimestamps configures Prometheus whether it should
```
Comment (bwplotka) [inline example/prometheus-operator-crd/prometheus.crd.yaml]: `overrideHonorTimestamps` in description?

Also can we elaborate more? This flag allows to set honor_timestamp per ServiceMonitor right? Can put this into description? (:

File: pkg/prometheus/promcfg_test.go [hunk 19]
```diff
@@ -1581,3 +1587,53 @@ func TestHonorLabels(t *testing.T) {
...
+		UserHonorTimestamps     *bool
+		OverrideHonorTimestamps bool
+		Expected                bool
+	}
+
+	no := false
+	yes := true
+	testCases := []testCase{
+		{
+			UserHonorTimestamps:     nil,
```
Comment (bwplotka) [inline pkg/prometheus/promcfg_test.go]: `User` means here `serviceMonitor` honor timestamp?

Comment (bwplotka) [reply pkg/prometheus/promcfg_test.go]: Can we be explicit?

File: pkg/prometheus/promcfg.go [hunk 20]
```diff
@@ -322,14 +322,29 @@ func honorLabels(userHonorLabels, overrideHonorLabels bool) bool {
 	return userHonorLabels
 }
 
+// honorTimestamps determines the value of honor_timestamps.
+// if overrideHonorTimestamps is true, then we want to always
+// set honor_timestamps to true. Otherwise we set it to whateever
+// user wants, while assuming no setting as true.
+func honorTimestamps(userHonorTimestamps *bool, overrideHonorTimestamps bool) bool {
```
Comment (bwplotka) [inline pkg/prometheus/promcfg.go]: I am so confused if `overrideHonorTimestamps` means `allowOverridingHonorTimestamp` or actually `overriddenHonorTimestamp` :see_no_evil: I think the latter.. based on comment.

File: Documentation/api.md (hunk 9 above)
Comment (paulfantom) [reply Documentation/api.md]: By default this should be `nil` to preserve backwards compatibility.

File: pkg/prometheus/promcfg.go (hunk 15 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg.go]: Agreed, but it is also not in scope of this PR :)

File: pkg/prometheus/promcfg_test.go (hunk 19 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: This is a convention from testing honor_labels, I didn't want to diverge.

File: pkg/prometheus/promcfg.go (hunk 20 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg.go]: `overrideHonorTimestamps: true` discards whatever was set in ServiceMonitor/PodMonitor and sets `honor_timestamps: true` in each scrape config.

File: Documentation/api.md (hunk 9 above)
Comment (paulfantom) [reply Documentation/api.md]: Also, this is `false` because generator made it this way :P

**Comment by paulfantom** [comment]
Addressed all comments.

**Comment by brancz** [comment]
CI failed. Not sure if legimately timed out or flake.

File: pkg/prometheus/promcfg_test.go [hunk 21]
```diff
@@ -1581,3 +1588,53 @@ func TestHonorLabels(t *testing.T) {
...
 }
+
+func TestHonorTimestamps(t *testing.T) {
+	type testCase struct {
+		UserHonorTimestamps     *bool
+		OverrideHonorTimestamps bool
+		Expected                bool
+	}
+
+	//no := false
```
Comment (LiliC) [inline pkg/prometheus/promcfg_test.go]: Are these leftovers from testing?

**Comment by paulfantom** [comment]
@brancz This is legitimate failure. We need to update required promehteus to 2.9.0 as this is a minimal version that implements `honor_timestamps` in `scrape_config`. Currently we are testing on 2.7.1

**Comment by paulfantom** [comment]
Seems like my previous fix wasn't good enough. Fixed it with revamped logic. Right now we should following truth table (override is horizontal, PodMonitor/ServiceMonitor is vertical):

|   | 0 | 1 |
|---|---|---|
| 0 | 0 | 1 |
| 1 | 1 | 1 |
| nil | nil/0 | 1 |

**Comment by paulfantom** [comment]
Back to the drawing board as I need to check some more things regarding this PR :)

File: pkg/prometheus/promcfg.go (hunk 15 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg.go]: fixed.

File: pkg/prometheus/promcfg_test.go (hunk 19 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: > User means here serviceMonitor honor timestamp?

ServiceMonitor or PodMonitor

**Comment by paulfantom** [comment]
For some reason, (resource exhaustion?) tests are failing in CI on testing FS access. If the test is not run in parallel, everything works.

**Comment by LiliC** [comment]
Hmm that's strange, I don't see the tests failing for other new PRs.

File: pkg/prometheus/promcfg.go [hunk 22]
```diff
@@ -341,6 +355,13 @@ func (cg *configGenerator) generatePodMonitorConfig(version semver.Version,
 		},
 	}
 
+	// If timestamps honoring features are not set, then don't enforce anything.
+	// This is here to preserve backwards compatibility with prometheus <2.9.0
+	if ep.HonorTimestamps != nil || overrideHonorTimestamps {
```
Comment (LiliC) [inline pkg/prometheus/promcfg.go]: I would just pass all the values to honorTimestamps and handle everything in that function.

File: pkg/prometheus/promcfg.go [hunk 23]
```diff
@@ -580,6 +602,13 @@ func (cg *configGenerator) generateServiceMonitorConfig(
...
```
Comment (LiliC) [inline pkg/prometheus/promcfg.go]: Same suggestion as above.

File: pkg/prometheus/promcfg_test.go [hunk 24]
```diff
@@ -465,6 +466,256 @@ alerting:
 	}
 }
 
+func TestSettingHonorTimestamps(t *testing.T) {
```
Comment (LiliC) [inline pkg/prometheus/promcfg_test.go]: As this is done for both PodMonitor and ServiceMonitor I would expect both to be tested. Here we are only testing ServiceMonitor.

File: pkg/prometheus/promcfg.go (hunk 22 above)
Comment (LiliC) [reply pkg/prometheus/promcfg.go]: This is mainly so the unit test you added covers these cases as well.

Comment (paulfantom) [reply pkg/prometheus/promcfg.go]: I didn't have any idea on how to easily do that as this function would need to return a tristate value. Two ways I could think of how this could be done the way you suggest is to either:
1) return two values
2) initialize a helper variable before executing a function and then passing a pointer to that variable as a function parameter.

The way it is written right now, `honorTimestamps` handles all cases for prometheus >=2.9.0 and this check is only to ensure compatibility with prometheus <2.9.0. Unit tests handle all 6 outcome values of `honorTimestamps`.

Both resulted in a more complicated codebase. If you have a better idea of how this can be handled, then I am open to suggestions.

File: pkg/prometheus/promcfg_test.go (hunk 24 above)
Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: Test added

Comment (LiliC) [reply pkg/prometheus/promcfg_test.go]: Sure, but thats outside of the scope of this PR.

File: pkg/prometheus/promcfg.go (hunk 22 above)
Comment (LiliC) [reply pkg/prometheus/promcfg.go]: Why can’t you pass in the cfg?

**Comment by paulfantom** [comment]
rebased

File: pkg/prometheus/promcfg.go [hunk 25]
```diff
@@ -322,12 +322,32 @@ func honorLabels(userHonorLabels, overrideHonorLabels bool) bool {
 	return userHonorLabels
 }
 
-func (cg *configGenerator) generatePodMonitorConfig(version semver.Version,
+// honorTimestamps adds option to enforce honor_timestamps option in scrape_config.
+// We want to disable honoring timestamps when user specified it or when global
+// override is set. For backwards compatibility with prometheus <2.9.0 we don't
```
[the remark this answers is not in the document]
Comment (brancz) [reply pkg/prometheus/promcfg.go]: Nevermind, it looks like we're doing that now outside of this function.

[the remark this answers is not in the document]
Comment (paulfantom) [reply pkg/prometheus/promcfg.go]: Added version protection in each `generate*MonitorConfig`. I didn't want to do it in honorTimestamps() as all version protections are in `generate*MonitorConfig` functions and it might cloud codebase a bit.

**Comment by brancz** [comment]
lgtm on green

File: pkg/prometheus/promcfg_test.go [hunk 26]
```diff
@@ -634,6 +635,369 @@ alerting:
...
+func TestSettingHonorTimestampsInServiceMonitor(t *testing.T) {
+	cg := &configGenerator{}
+	cfg, err := cg.generateConfig(
+		&monitoringv1.Prometheus{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test",
+				Namespace: "default",
+			},
+			Spec: monitoringv1.PrometheusSpec{
+				Version: "v2.9.0",
```
Comment (LiliC) [inline pkg/prometheus/promcfg_test.go]: I don't remember this needing to be here when I reviewed last week? 🤔

Comment (paulfantom) [reply pkg/prometheus/promcfg_test.go]: This is now necessary due to https://github.com/coreos/prometheus-operator/pull/2800#discussion_r336875968
